File: tests/fuzzy_match/fts_fuzzy_match_test.cpp
// LICENSE
//
//   This software is dual-licensed to the public domain and under the following
//   license: you are granted a perpetual, irrevocable license to copy, modify,
//   publish, and distribute this file as you see fit.

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <unordered_map>
#include <queue>

#include "..\..\code\fts_fuzzy_match.h"
#include "..\..\code\util\fts_timer.h"


int main(int argc, char *argv[]) {

    // Dictionary
    std::vector<std::string> dictionary;
    
    auto countMatches = [&dictionary](std::string const & pattern) -> int { 
        int matches = 0;
        for (auto && entry : dictionary)
            if (fts::fuzzy_match(pattern.c_str(), entry.c_str()) == 1)
                ++matches;

        return matches; 
    };

    auto alphabeticalMatches = [&dictionary](std::string const & pattern) {
        std::vector<std::string const *> matches;
        for (auto && entry : dictionary)
            if (fts::fuzzy_match(pattern.c_str(), entry.c_str()) == 1)
                matches.push_back(&entry);

        return matches;
    };

    auto scoredMatches = [&dictionary](std::string const & pattern) {
        std::vector<std::pair<int, std::string const*>> matches;
        int score;
        for (auto && entry : dictionary)
            if (fts::fuzzy_match(pattern.c_str(), entry.c_str(), score))
                matches.emplace_back(score, &entry);

        std::sort(matches.begin(), matches.end(), [](auto && a, auto && b) { return a.first > b.first; });
        
        return matches;
    };

    // Open file
    using namespace std::string_literals;
    std::string path = argc > 1 ? argv[1] : "no file specified"s;
    std::cout << "Reading [" << path << "]" << std::endl;
    std::ifstream infile(path);
    if (!infile.good()) {
        std::cout << "Failed to open file." << std::endl;
        return 0;
    }

    // Read file
    fts::Stopwatch stopwatch;
    std::string entry;
    while (std::getline(infile, entry))
        dictionary.push_back(std::move(entry));

    float time = stopwatch.elapsedMilliseconds();
    std::cout << "Read [" << dictionary.size() << "] entries in " << time << "ms" << std::endl << std::endl;


    // Input Loop
    std::string option;
    std::string pattern;

    while (1) {
        std::cout << "1. Count Matches" << std::endl;
        std::cout << "2. Print Matches (Alphabetical)" << std::endl;
        std::cout << "3. Print Matches (By Score)" << std::endl;
        std::cout << "4. Exit" << std::endl << std::endl;
        std::cout << "> ";
        std::getline(std::cin, option);
        std::cout << std::endl;

        if (option == "1" || option == "2" || option == "3") {

            // Read pattern from std::cin
            std::cout << "Enter search pattern" << std::endl << std::endl << "> ";
            std::getline(std::cin, pattern);
            std::cout << std::endl << std::endl << "Searching for [" << pattern << "] ..." << std::endl;

            if (option == "1") {
                // Count Matches
                stopwatch.Reset();
                int matches = countMatches(pattern);
                time = stopwatch.elapsedMilliseconds();

                std::cout << "Found " << matches << " matches in " << time << "ms" << std::endl << std::endl;
            }
            else if (option == "2") {
                // Print Matches (Alphabetical)
                stopwatch.Reset();
                auto results = alphabeticalMatches(pattern);
                time = stopwatch.elapsedMilliseconds();

                for (auto && result : results)
                    std::cout << *result << std::endl;
                std::cout << std::endl << "Found " << results.size() << " matches in " << time << "ms" << std::endl << std::endl;
            }
            else if (option == "3") {
                // Print Matches (By Score)
                stopwatch.Reset();
                auto results = scoredMatches(pattern);
                time = stopwatch.elapsedMilliseconds();

                for (auto && result : results)
                    std::cout << result.first << " - " << *result.second << std::endl;
                std::cout << std::endl << "Found " << results.size() << " matches in " << time << "ms" << std::endl << std::endl;
            }
        }
        else if (option == "4") {
            // Quit
            break;
        }
    }
}

File: tests/fuzzy_match/fts_fuzzy_match_test.cs
// LICENSE
//
//   This software is dual-licensed to the public domain and under the following
//   license: you are granted a perpetual, irrevocable license to copy, modify,
//   publish, and distribute this file as you see fit.

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using fts;

namespace FuzzyMatch
{
	internal class Program
	{
		private static string[] _dictionary;

		private static void Main(string[] args)
		{
			if (args.Length != 1)
			{
				Console.WriteLine("No file specified.");
				return;
			}
			var path = args[0];
			Console.WriteLine($"Reading [{path}]");

			// Open and read file
			var watch = Stopwatch.StartNew();
			_dictionary = File.ReadAllLines(path);
			watch.Stop();

			var time = watch.Elapsed.TotalMilliseconds;
			Console.WriteLine($"Read [{_dictionary.Length}] entries in {time} ms.");

			// Input Loop
			while (true)
			{
				Console.WriteLine("1. Count Matches");
				Console.WriteLine("2. Print Matches (Alphabetical)");
				Console.WriteLine("3. Print Matches (By Score)");
				Console.WriteLine("4. Exit");
				Console.WriteLine();
				Console.Write("> ");
				var option = Console.ReadLine();
				Console.WriteLine();

				if (option == "1" || option == "2" || option == "3")
				{
					// Read pattern
					Console.WriteLine("Enter search pattern");
					Console.WriteLine();
					Console.Write("> ");
					var pattern = Console.ReadLine();
					Console.WriteLine();
					Console.WriteLine();
					Console.WriteLine($"Searching for [{pattern}] ...");
					Console.WriteLine();
					if (option == "1")
					{
						// Count Matches
						watch.Reset();
						watch.Start();
						var matches = CountMatches(pattern);
						time = watch.Elapsed.TotalMilliseconds;

						Console.WriteLine($"Found {matches} matches in {time} ms.");
						Console.WriteLine();
						Console.WriteLine();

					}
					else if (option == "2")
					{
						// Print Matches (Alphabetical)
						watch.Reset();
						watch.Start();
						var results = AlphabeticalMatches(pattern);
						time = watch.Elapsed.TotalMilliseconds;

						foreach (var result in results)
							Console.WriteLine(result);

						Console.WriteLine();
						Console.WriteLine($"Found {results.Count} matches in {time} ms.");
						Console.WriteLine();
						Console.WriteLine();

					}
					else if (option == "3")
					{
						// Print Matches (By Score)
						watch.Reset();
						watch.Start();
						var results = ScoredMatches(pattern);
						time = watch.Elapsed.TotalMilliseconds;

						foreach (var result in results)
							Console.WriteLine($"{result.Key} - {result.Value}");

						Console.WriteLine();
						Console.WriteLine($"Found {results.Count} matches in {time} ms.");
						Console.WriteLine();
						Console.WriteLine();
					}
				}
				else if (option == "4")
				{
					// Quit
					break;
				}
			}
		}

		private static int CountMatches(string pattern)
		{
			var matches = 0;
			foreach (var entry in _dictionary)
				if (FuzzyMatcher.FuzzyMatch(pattern, entry))
					++matches;

			return matches;
		}

		private static List<string> AlphabeticalMatches(string pattern)
		{
			var matches = new List<string>();
			foreach (var entry in _dictionary)
				if (FuzzyMatcher.FuzzyMatch(pattern, entry))
					matches.Add(entry);

			return matches;
		}

		private static Dictionary<string, int> ScoredMatches(string pattern)
		{
			var matches = new List<KeyValuePair<string, int>>();
			int score;
			foreach (var entry in _dictionary)
				if (FuzzyMatcher.FuzzyMatch(pattern, entry, out score))
					matches.Add(new KeyValuePair<string, int>(entry, score));

			return matches.OrderBy(item => -item.Value).ToDictionary(item => item.Key, item => item.Value);
		} 
	}
}
